habitat/core/env.py

```diff
@@ -423,3 +425,276 @@ def __enter__(self):
+                prob = self._config.CHANGE_TASK_BEHAVIOUR.get(
+                    "CHANGE_TASK_PROB", 0.5
+                )
+                if (
+                    eps_change
+                    and self._eps_counter >= eps_change
+                    and np.random.choice([1, 0], 1, p=[prob, 1 - prob]).item()
+                ):
+                    change_ = True
+                elif (
+                    eps_change is None
+                    and cum_steps_change is None
+                    and self._eps_counter >= eps_change
+                    and np.random.choice([1, 0], 1, p=[prob, 1 - prob]).item()
+                ):
+                    change_ = True
+
+            if change_:
+                self._eps_counter = 0
+        elif cum_steps_change:
+            # episode is still on-going
+            if self._change_task_behavior == CHANGE_TASK_BEHAVIOUR.FIXED.name:
+                if self._cumulative_steps_counter >= cum_steps_change:
+                    change_ = True
+            elif (
+                self._change_task_behavior == CHANGE_TASK_BEHAVIOUR.RANDOM.name
+            ):
+                # change task with some prob if we exceed min number of steps
+                prob = self._config.CHANGE_TASK_BEHAVIOUR.get(
+                    "CHANGE_TASK_PROB", 0.5
+                )
+                if (
+                    self._cumulative_steps_counter >= cum_steps_change
+                    and np.random.choice([1, 0], 1, p=[prob, 1 - prob]).item()
+                ):
+                    change_ = True
+
+            if change_:
+                self._cumulative_steps_counter = 0
+        return change_
+
+    def _change_task(self, is_reset=False):
+        """Change current task according to specified loop behaviour.
+
+        Args:
+            is_reset (bool, optional): [description]. Defaults to False.
+        """
+        prev_task_idx = self._curr_task_idx
+        if (
+            self._task_cycling_behavior
+            == CHANGE_TASK_CYCLE_BEHAVIOUR.ORDER.name
+        ):
+            self._curr_task_idx = (self._curr_task_idx + 1) % len(self._tasks)
+        elif (
+            self._task_cycling_behavior
+            == CHANGE_TASK_CYCLE_BEHAVIOUR.RANDOM.name
+        ):
+            # sample from other tasks with equal probability
+            self._curr_task_idx = np.random.choice(
+                [
+                    i
+                    for i in range(len(self._tasks))
+                    if i != self._curr_task_idx
+                ],
+                1,
+            ).item()
+
+        self._task = self._tasks[self._curr_task_idx]
+        # update episode iterator with task dataset
+        iter_option_dict = {
+            k.lower(): v
+            for k, v in self._config.ENVIRONMENT.ITERATOR_OPTIONS.items()
+        }
+        iter_option_dict["seed"] = self._config.SEED
+        self._episode_iterator: EpisodeIterator = (
```

Comment:
When the task A was switched and after the loop switched to Task A again, does that mean that iteration will start from the initial state as EpisodeIterator is recreated each time? That means that there can be part of episode dataset that will never be reached. Maybe that make more sense to keep iterators alive for each task?

Reply:
Yeah probably keeping iterators alive would provide a more consistent experience too. I'll implement that.